<!-- 
  Simulate a differential drive robot with the teb_local_planner in gazebo:
  - gazebo
  - map_server
  - move_base
  - static map
  - fake localization
  - rviz view
 -->
<launch>

  <!--  ************** Global Parameters ***************  -->
  <param name="/use_sim_time" value="true"/>
  <arg name="gui" default="true"/>

  <!-- Initial pose -->
  <arg name="initial_pose_x" default="2.0"/>
  <arg name="initial_pose_y" default="0.0"/>
  <arg name="initial_pose_a" default="1.5708"/>

  <!-- Filename -->
  <arg name="log_filename" default="$(find move_base_benchmark)/launch/log.txt" />
  <arg name="map_filename" default="$(find move_base_benchmark)/maps/small_house/map.yaml" />
  <arg name="world_filename" default="$(find gazebo_world)/world/small_house_map.world" />
  <arg name="global_frame" value="map" />

  <!--  ************** Navigation *************** -->
  <node pkg="move_base_benchmark" type="move_base_benchmark" respawn="false" name="move_base_benchmark" output="screen">
    <rosparam file="$(find move_base_benchmark)/params/costmap_common_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find move_base_benchmark)/params/costmap_common_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find move_base_benchmark)/params/global_costmap_params.yaml" command="load" />
    <rosparam file="$(find move_base_benchmark)/params/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find move_base_benchmark)/params/teb_local_planner_params.yaml" command="load" />
    <param name="global_costmap/global_frame" value="$(arg global_frame)" />
    <param name="local_costmap/global_frame" value="$(arg global_frame)" />

    <param name="base_global_planner" value="global_planner/GlobalPlanner" />
    <param name="planner_frequency" value="5.0" />
    <param name="planner_patience" value="5.0" />

    <param name="base_local_planner" value="teb_local_planner/TebLocalPlannerROS" />
    <param name="controller_frequency" value="15.0" />
    <param name="controller_patience" value="15.0" />

    <param name="log_filename" value="$(arg log_filename)" />
    <remap from="cmd_vel" to="RosAria/cmd_vel"/>
    <remap from="odom" to="RosAria/odom"/>
  </node> 
  <!--  ****** Maps *****  -->
  <node name="map_server" pkg="map_server" type="map_server" args="$(arg map_filename)" output="screen">
    <param name="frame_id" value="$(arg global_frame)"/>
  </node>

  <!--  ****** Localization *****  -->
  <node name="fake_localization" pkg="fake_localization" type="fake_localization" respawn="false" >
    <param name="global_frame_id" value="$(arg global_frame)" />
  </node>

  <!--  ****** Start Gazebo and launch world model *****  -->
  <include file="$(find gazebo_world)/launch/world_launch.launch">
    <arg name="world_name" value="$(arg world_filename)"/>
    <arg name="gui" value="$(arg gui)" />
  </include>

  <!-- Spawn p3dx mobile robot -->
  <include file="$(find p3dx_gazebo)/launch/spawn_p3dx.launch">
    <arg name="initial_pose_x" value="$(arg initial_pose_x)" />
    <arg name="initial_pose_y" value="$(arg initial_pose_y)" />
    <arg name="initial_pose_a" value="$(arg initial_pose_a)" />
  </include>

  <!-- Spawn segway mobile robot 
  <include file="$(find segway_gazebo)/launch/spawn_segway.launch"/>-->
  <!-- Launch widowx gazebo ros_control 
  <include file="$(find widowx_turret_controller)/launch/widowx_turret_gazebo_controller.launch"/>-->

  <!--  **************** Visualisation ****************  -->
  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find move_base_benchmark)/launch/rviz_navigation.rviz"/>

  <!--  ****** Actor Position Publisher *****  -->
  <node name="actor_pos_publish_node" pkg="actor_pos_publish" type="actor_pos_publish_node" respawn="true" 
  launch-prefix="bash -c 'sleep 8; $0 $@' "/>
  
</launch>